viadot/sources/salesforce.py

```diff
@@ -0,0 +1,238 @@
+        raise_on_error: bool = False,
+    ) -> None:
+        """
+        Performs a Bulk upsert to Salesforce of the data given in the Dataframe.
+        Bulk upsert is performed on multiple records simultaneously, it is usually used when
+        there is a need to insert or update multiple records in a single transaction,
+        which can be more efficient and reduce the number of API calls required.
+
+        Args:
+            df (pd.DataFrame): Dataframe containing the rows to Bulk upsert.
+            table (str): The table where the data should be upserted.
+            external_id_column (str, optional): The external ID to use for the upsert. Defaults to None.
+            batch_size (int, optional): Number of records to be included in each batch of records
+                that are sent to the Salesforce API for processing. Defaults to 10000.
+            raise_on_error (bool, optional): Whether to raise an exception if a row upsert fails.
+                If False, we only display a warning. Defaults to False.
+        """
+        if df.empty:
+            self.logger.info("No data to upsert.")
+            return
+
+        if external_id_column and external_id_column not in df.columns:
+            raise ValueError(
+                f"Passed DataFrame does not contain column '{external_id_column}'."
+            )
+        records = df.to_dict("records")
+
+        try:
+            response_code = self.salesforce.bulk.__getattr__(table).upsert(
+                data=records,
+                external_id_field=external_id_column,
+                batch_size=batch_size,
+            )
+        except SalesforceMalformedRequest as e:
+            # Bulk insert didn't work at all.
+            raise ValueError(f"Upsert of records failed: {e}") from e
+
+        self.logger.info(f"Successfully upserted bulk records.")
+
+        if any(result.get("success") is not True for result in response_code):
+            # Upsert of some individual records failed.
+            failed_records = [
+                result for result in response_code if result.get("success") is not True
+            ]
+            msg = f"Upsert failed for records {failed_records} with response code {response_code}"
+            if raise_on_error:
+                raise ValueError(msg)
+            else:
+                self.logger.warning(msg)
+
+        self.logger.info(
+            f"Successfully upserted {len(records)} records into table '{table}'."
+        )
+
+    def download(
+        self, query: str = None, table: str = None, columns: List[str] = None
+    ) -> List[OrderedDict]:
+        """
+        Download all data from the indicated table or the result of the specified query.
+
+        Args:
+            query (str, optional): Query for download the specific data. Defaults to None.
+            table (str, optional): Table name. Defaults to None.
+            columns (List[str], optional): List of columns which are needed,
```

Comment:
"List of columns which are needed, requires table argument"-> "list of required columns; requires `table` to be specified" (appears 2x)

Reply:
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)